src/pipx/commands/upgrade.py

```diff
@@ -28,6 +28,15 @@ def _upgrade_package(
 
     if package_metadata.package_or_url is None:
         raise PipxError(f"Internal Error: package {package_name} has corrupt pipx metadata.")
+    elif package_metadata.pinned and package_metadata.package != venv.main_package_name:
+        logger.warning(
+            f"Not upgrading pinned package {package_metadata.package} in venv {venv.name}. "
+            f"Run `pipx unpin {venv.name}` to unpin it."
+        )
+        return 0
+    elif package_metadata.pinned:
+        logger.warning(f"Not upgrading pinned package {venv.name}. Run `pipx unpin {venv.name}` to unpin it.")
+        return 0
```

Comment:
I wonder if this should return an error instead.

Reply:
Returning 0 here means the version of package is not changed, and raising error may terminate the upgrade-all process.

Reply:
From my perspective, returning an error here would defeat the purpose of this feature, which is to be able to run `pipx upgrade-all` to upgrade all packages _except_ pinned packages — without errors and with return code of `0`. In terms of prior art, Homebrew emits warnings in these cases:

```fish
~ ➤ brew upgrade --formula

Warning: Not upgrading 1 pinned package:
foo-package 1.7.2
```